<h1> Case Study Submission </h1>
<h4> Course: 040669, Simulation 1</h4>
<h4> Date: 25.11.2024 </h4>
<h4> Name: Jakob Lindstrøm </h4>
<br>
<h2> i) Solution description </h2>
<p> I have solved case 2: Doctor's office. The answers to the subquestions of the case will be solved using programming languages. The answers will be written in this notebook, but some of the answers might be derived using the following repo: <a href="https://github.com/DataJakob/040669-Simulations-1-and-2/tree/main/RD_que_system"> DataJakob@github</a>.</p>
<br>

<h2> i) Reproduce the solution </h2>
<p> In IDE terminal -> git clone repo </p>
<p> Navigate to -> deliverable.ipynbd </p>
<p> Choose interpreter -> Julia  </p>
<p> Press ->  Run all </p>
<br>

<h2> iii)  General Solution Comments </h2>
<p> For this task I have choosen to use both an analytical and numerical approach to retrieve the key parameters for the first subtask.  </p>

In [ ]:
include("../Formulas/MM.jl")
include("../RD_que_system/src/00_arriver.jl")
include("../RD_que_system/src/01_server.jl")
include("../RD_que_system/src/02_run_queue.jl")
include("../RD_que_system/src/03_calculate_parameters.jl")

calculate_MMS (generic function with 1 method)

<h2> Task 2a </h2>
<p> To solve this task I leverage a class and methods from the the file MM.jl. Doing this allows me to calculate the key parameters for both systems. </p>

In [ ]:
n_lines = 1               # number of lines
n_time = 8.0 *5*4          # 8h * 5d * 4.2w *12m
n_arr_hour = 4.0          # arrivals per hour
n_ser_hour = 5.0          # servie per hour
n_servers = 1             # Number of servers
n_packages_queue_overload = 10000   # When to switch queue line

wait_cum_MM1 = []
length_cum_MM1 =  []
for i in 1:1000
    randomness = i

    a,b,c = calculate_ia_and_s(n_lines, n_time, n_arr_hour, n_ser_hour, randomness) # arr | cum_arr | ser
    my_server = ServerBase(n_servers, n_packages_queue_overload)
    
    Data_to_json(a, c, b)
    Run_the_queue(my_server, 100000.0)
    wait_time, queue_length = Calculate_key_parameters(Int(n_time*n_arr_hour))
    push!(wait_cum_MM1, wait_time[1])
    push!(length_cum_MM1, queue_length[1])
end


MM1 = queueKeyParameters(n_arr_hour, n_ser_hour, 0)
MM1 = calcuclate_MM1(MM1)

In [221]:
println("Average length in system for MM1: ", MM1.L)
println("Average time in system for MM1: ", MM1.W)
println("\n\tCompare parameters for Lq and Wq using analytical and simulation approach")
println("Analytical:")
println("Mean wait time: ",MM1.Wq)
println("Mean queue length: ",MM1.Lq)
println("\nSimulation:")
println("Mean wait time: ",round(mean(wait_cum_MM1), digits=2))
println("Mean queue length: ",round(mean(length_cum_MM1), digits=2))

Average length in system for MM1: 4.0
Average time in system for MM1: 1.0

	Compare parameters for Lq and Wq using analytical and simulation approach
Analytical:
Mean wait time: 0.8
Mean queue length: 3.2

Simulation:
Mean wait time: 0.76
Mean queue length: 3.06


In [223]:
MMS = queueKeyParameters(4.0, 2.5, 0)
MMS = calculate_MMS(MMS, 2)


println("\nAverage length in system for MM2: ", MMS.L)
println("Average time in system for MM2: ", MMS.W)


Average length in system for MM2: 2.2130268199233716
Average time in system for MM2: 0.5532567049808429


In [224]:
println("Analytical:")
println("Mean wait time: ",MMS.Wq)
println("Mean queue length: ",MMS.Lq)
println("\nSimulation:")
println("Mean wait time: ",round(mean(alfa), digits=2))
println("Mean queue length: ",round(mean(bravo), digits=2))

Analytical:
Mean wait time: 0.1532567049808429
Mean queue length: 0.6130268199233716

Simulation:
Mean wait time: 0.65
Mean queue length: 2.58


<b> Comments: </b>
<p> Both of the systems have the same lambda and mu. The MM2 system that can handle two customers simultaneously and has a significantly higher queue lengths and waiting times in the system compared to the MM1 system. </p>
<br>
<p> I believe this difference is a result of longer processing times for the servers. The  servers are twice as slow compared to the MM1 server, but there are twice as many servers in contrast. </p>
<br>

<h2> Task 2b </h2>
<p> In order to make the recommendation for Caroline's new administrative system I use the cost and the key parameters from the respective queues.

In [3]:
println("Hourly costs of MM1: ", 14.8)
println("Hourly costs of MM2: ", 13.2)

Hourly costs of MM1: 14.8
Hourly costs of MM2: 13.2


<p> I would recommend for Caroline to choose the alternative "Healthy in No Time"; the MM2 system. The reason for this is that the system has a lower costs per hour, even though the average length and average waiting time in the system is longer. </p>
<br>
<p> From a business point of perspective I would focus on revenues and costs, and therefore, deprioritize the customers's experience. This recommendation stands, until other information reveals itself.</p>

<h2> Task 2c - i </h2>
<p> Given the new information about the fine issued from the health insurance of the city, I need to calculate the costs for having patitents in line. </p>
<br>
<p> In order to solve this task I would calculate the amount of minutes each customer must wait during an hour multiply that with the average number of customers in a line and then multiplied that with the costs of having customers in line. 

In [4]:
println("Cost for MM2: ",round(MMS.Wq*60*MMS.Lq*3.45))

Cost for MM2: 775.0


<p> The hourly cost of the new adminsitrative system, given the administrative cost and the fines are 775+13.2 = 788.2. </p>¨
<br>
<h3> 2c - ii, Alternative recommendation </h3>
<p> Based on these results I would do the calculations for the second alternative and might reevaluate my  recommendation. Another recommendation I could propose to Caroline is to implement a finite queue. This would probably reduce here probability of having max a full queue at all times, but it would proabbly significantly lower her costs for having customers in queue. 

In [5]:
println("Cost for MM1: ",round(MM1.Wq*60*MM1.Lq*3.45))

Cost for MM1: 530.0


<p> The second alternative, the MM1 seems to be a lot better in terms of costs saved, because the costs are significantly lower. 

In [6]:
include("../RD_que_system/src/00_arriver.jl")
include("../RD_que_system/src/01_server.jl")
include("../RD_que_system/src/02_run_queue.jl")
include("../RD_que_system/src/03_calculate_parameters.jl")

Calculate_key_parameters (generic function with 1 method)

In [7]:
using Distributions

In [185]:
function calculate_ia_and_s(
    n_line::Int64,
    n_time::Float64,
    n_arr_hour::Float64,
    n_ser_hour::Float64,
    randomness::Int64,
    )

    # n_arr_rate = n_arr_hour/60
    arrival =  Vector{Vector{Tuple{String, Float64}}}()
    cum_arrival = Vector{Vector{Tuple{String, Float64}}}()
    service = Vector{Vector{Tuple{String, Float64}}}()

    for i in 1:n_line
        Random.seed!(randomness+i)
        single_arrival = rand(Exponential(1/n_arr_hour), Int(n_time*n_arr_hour))
        single_cum_arrival = cumsum(single_arrival)
        single_service = rand(Exponential(1/n_ser_hour), Int(n_time*n_arr_hour))

        single_arrival_with_ids = [
            (string(Char('a' + i - 1)) * string(j), single_arrival[j]) for j in 1:length(single_arrival)
        ]

        # Create DataEntry for cumulative arrival times with unique IDs
        single_cum_arrival_with_ids = [
            (string(Char('a' + i - 1)) * string(j), single_cum_arrival[j]) for j in 1:length(single_cum_arrival)
        ]

        # Create DataEntry for service times with unique IDs
        single_service_with_ids = [
            (string(Char('a' + i - 1)) * string(j), single_service[j]) for j in 1:length(single_service)
        ]

        # Store these in the respective vectors
        push!(arrival, single_arrival_with_ids)
        push!(cum_arrival, single_cum_arrival_with_ids)
        push!(service, single_service_with_ids)
    end

    return [arrival, cum_arrival, service]
end

calculate_ia_and_s (generic function with 1 method)

In [ ]:
n_lines = 1               # number of lines
n_time = 8.0 *5*4          # 8h * 5d * 4.2w *12m
n_arr_hour = 4.0          # arrivals per hour
n_ser_hour = 5.0          # servie per hour
n_servers = 1             # Number of servers
n_packages_queue_overload = 10000   # When to switch queue line

wait_cum_MM1 = []
length_cum_MM1 =  []
for i in 1:1000
    randomness = i

    a,b,c = calculate_ia_and_s(n_lines, n_time, n_arr_hour, n_ser_hour, randomness) # arr | cum_arr | ser
    my_server = ServerBase(n_servers, n_packages_queue_overload)
    
    Data_to_json(a, c, b)
    Run_the_queue(my_server, 100000.0)
    wait_time, queue_length = Calculate_key_parameters(Int(n_time*n_arr_hour))
    push!(wait_cum_MM1, wait_time[1])
    push!(length_cum_MM1, queue_length[1])
end

In [218]:
println("Analytical:")
println("Mean wait time: ",MM1.Wq)
println("Mean queue length: ",MM1.Lq)
println("\nSimulation:")
println("Mean wait time: ",round(mean(wait_cum_MM1), digits=2))
println("Mean queue length: ",round(mean(length_cum_MM1), digits=2))

Analytical:
Mean wait time: 0.8
Mean queue length: 3.2

Simulation:
Mean wait time: 0.75
Mean queue length: 3.03


In [ ]:
n_lines = 1               # number of lines
n_time = 8.0 *5*4          # 8h * 5d * 4.2w *12m
n_arr_hour = 4.0          # arrivals per hour
n_ser_hour = 2.5          # servie per hour
n_servers = 2             # Number of servers
n_packages_queue_overload = 10000   # When to switch queue line

alfa = []
bravo =  []
for i in 1:1000
    randomness = i

    a,b,c = calculate_ia_and_s(n_lines, n_time, n_arr_hour, n_ser_hour, randomness) # arr | cum_arr | ser
    my_server = ServerBase(n_servers, n_packages_queue_overload)
    
    Data_to_json(a, c, b)
    Run_the_queue(my_server, n_time*1.5)
    wait_time, queue_length = Calculate_key_parameters(Int(n_time*n_arr_hour))
    push!(alfa, wait_time[1])
    push!(bravo, queue_length[1])
end

In [216]:
println("Analytical:")
println("Mean wait time: ",MMS.Wq)
println("Mean queue length: ",MMS.Lq)
println("\nSimulation:")
println("Mean wait time: ",round(mean(alfa), digits=2))
println("Mean queue length: ",round(mean(bravo), digits=2))

Analytical:
Mean wait time: 0.9677419354838714
Mean queue length: 3.8709677419354858

Simulation:
Mean wait time: 0.65
Mean queue length: 2.58
